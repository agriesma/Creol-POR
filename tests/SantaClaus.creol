interface SantaClausR
begin with ReinDeer
  op backFromHoliday
end

interface SantaClausE
begin with Elf
  op haveProblem
end

interface SantaClaus inherits SantaClausR, SantaClausE
begin
end

interface Reindeer
begin with SantaClausR
  op harness
  op unharness
end

interface Elf
begin with SantaClausE
  op enterOffice
  op leaveOffice
end



class SantaClausC implements SantaClaus
begin
  var ct_reindeers:nat := 0
  var wait_reindeers:List := nil
  var harnessed_reindeers:List := nil
  var ct_elves:nat := 0
  var wait_elves:List := nil
  var inoffice_elves:List := nil

  op run == !loop()

  op loop == (await ct_reindeers = 5; deliverToys() []
              await ct_elves >= 3 and not ct_reindeers = 9; talkToElves()); 
             !loop()

  op deliverToys ==
	var t1:Label;
  	var t2: Label;
  	var t3: Label;
  	var t4: Label;
  	var t5:Label;
	ct_reindeers := 0;
	t1!first(wait_reindeers).harness(); 
	harnessed_reindeers := harnessed_reindeers :: first(wait_reindeers); 
		wait_reindeers := rest(wait_reindeers);
	-- ....;
		t5!first(wait_reindeers).harness(); 
		harnessed_reindeers := harnessed_reindeers :: first(wait_reindeers); 
	wait_reindeers := rest(wait_reindeers);
	await t1? and t2? and t3? and t4? and t5?;
	-- <Deliver Toys>;
	!first(harnessed_reindeers).unharness();
	harnessed_reindeers := rest(harnessed_reindeers);
	-- ....;
	!first(harnessed_reindeers).unharness();
	harnessed_reindeers := rest(harnessed_reindeers)


	op talkToelves ==
		var t1:Label;
		var t2:Label;
		var t3:Label;
		ct_elves := ct_elves - 3; 
		t1!first(wait_elves).showIn();
		inoffice_elves := inoffice_elves :: first(wait_elves);
		wait_elves := rest(wait_elves);
		t2!first(wait_elves).showIn();
		inoffice_elves := inoffice_elves :: first(wait_elves);
		wait_elves := rest(wait_elves);
		t3!first(wait_elves).showIn();
		inoffice_elves := inoffice_elves :: first(wait_elves);
		wait_elves := rest(wait_elves);
		await t1? and t2? and t3?;
		-- <Talk to elves>;
		t1!first(inoffcie_elves).showOut();
		inoffice_elves := rest(inoffice_elves);
		t2!first(inoffcie_elves).showOut();
		inoffice_elves := rest(inoffice_elves);
		t3!first(inoffcie_elves).showOut();
		inoffice_elves := rest(inoffice_elves);
		await t1? and t2? and t3?

with Reindeer
	op backFromHoliday ==
		ct_reindeers := ct_reindeers + 1; 
		wait_reindeers := wait_reindeers :: caller

with Elf
	op haveProblem ==
		ct_elves := ct_elves + 1; 
		wait_elves := wait_elves :: caller
end


class ReindeerC(sc:toReinDeer) implements Reindeer
begin
	op run == !holiday()
	op holiday == -- <go on holiday>;
		!sc.backFromHoliday()
	op deliverToys == skip -- <deliver Toys>

  with SantaClausR
	op harness == !deliverToys()
	op unharness == !holiday()
end



class ElfC(sc:SantaClaus) implements Elf
begin 
	op run == !work()
	op work == -- <do work>;
		!sc.haveProblem()
	op talkToSanta == skip -- <talk to Santa>

  with SantaClausE
	op showIn == !talkToSanta()
	op showOut == !work()
end

class LeaderElfC(sc:SantaClaus, elves:List)
     implements Elf inherits ElfC
begin
  op run == run@ElfC()
  op work ==
	-- (<Lead elves> ||| <Make toys>);
	!sc.haveProblem()
end

class Christmas
begin
  op run == 
	var sc:SantaC;
	var r:Reindeer;
	var e:Elf;
	sc := new SantaClausClass;
	r := new Reindeer(sc);
	r := new Reindeer(sc);
	r := new Reindeer(sc);
	r := new Reindeer(sc);
	r := new Reindeer(sc);
	e := new Elf(sc);
	e := new Elf(sc);
	e := new Elf(sc);
	e := new Elf(sc);
	e := new Elf(sc);
	e := new Elf(sc);
	e := new Elf(sc)
end

--	#COMMENTS:
--
--	Local synchronous calls: m(...;...) is same as t!m(...;..); t?() ---
--	an abbrevation. 
--
--	!m(..)  or with no arguments  !m  starts the method call m but does
--	not wait for any answer...
