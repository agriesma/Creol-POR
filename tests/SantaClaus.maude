load creol-interpreter
mod PROGRAM is
pr CREOL-INTERPRETER .
op init : -> Configuration [ctor] .
eq init =
< "SantaClausC" : Cl | Inh: noInh, Par: noAid, Att: "ctRreindeers" |-> null , "waitRreindeers" |-> null , "harnessedReindeers" |-> null , "ctElves" |-> null , "waitElves" |-> null , "inofficeElves" |-> null, Mtds: 
  < "run" : Mtdname | Param: noAid, Latt: noSubst, Code: ( ".anon" ! "this" . "loop" (emp ) )  ; free( ".anon" ) ; return ( emp ) > *
  < "loop" : Mtdname | Param: noAid, Latt: noSubst, Code: ( ( await "=" ( "ctReindeers" # int(5) ) ) ; ( "sync.1" ! "this" . "deliverToys" (emp ) )  ; ( "sync.1" ? ( noAid ) )  [] ( await "&&" ( "<=" ( int(3) # "ctElves" ) # "~" ( "=" ( "ctReindeers" # int(9) ) ) ) ) ; ( "sync.0" ! "this" . "talkToElves" (emp ) )  ; ( "sync.0" ? ( noAid ) )  ) ; ( ".anon" ! "this" . "loop" (emp ) )  ; free( ".anon" ) ; return ( emp ) > *
  < "deliverToys" : Mtdname | Param: noAid, Latt: noSubst, Code: "ctReindeers" ::= int(0) ; "t1" ! "first" ( "waitReindeers" ) . "harness" ( emp ) ; "harnessedReindeers" ::= "plus" ( "harnessedReindeers" # "first" ( "waitReindeers" ) ) ; "waitReindeers" ::= "rest" ( "waitReindeers" ) ; "t5" ! "first" ( "waitReindeers" ) . "harness" ( emp ) ; "harnessedReindeers" ::= "plus" ( "harnessedReindeers" # "first" ( "waitReindeers" ) ) ; "waitReindeers" ::= "rest" ( "waitReindeers" ) ; ( await ( "t1" ?? ) ) ; ( await ( "t2" ?? ) ) ; ( await ( "t3" ?? ) ) ; ( await ( "t4" ?? ) ) ; ( await ( "t5" ?? ) ) ; ".anon" ! "first" ( "harnessedReindeers" ) . "unharness" ( emp ) ; free( ".anon" ) ; "harnessedReindeers" ::= "rest" ( "harnessedReindeers" ) ; ".anon" ! "first" ( "harnessedReindeers" ) . "unharness" ( emp ) ; free( ".anon" ) ; "harnessedReindeers" ::= "rest" ( "harnessedReindeers" ) ; return ( emp ) > *
  < "talkToelves" : Mtdname | Param: noAid, Latt: noSubst, Code: "ctElves" ::= "-" ( "ctElves" # int(3) ) ; "t1" ! "first" ( "waitElves" ) . "showIn" ( emp ) ; "inofficeElves" ::= "plus" ( "inofficeElves" # "first" ( "waitElves" ) ) ; "waitElves" ::= "rest" ( "waitElves" ) ; "t2" ! "first" ( "waitElves" ) . "showIn" ( emp ) ; "inofficeElves" ::= "plus" ( "inofficeElves" # "first" ( "waitElves" ) ) ; "waitElves" ::= "rest" ( "waitElves" ) ; "t3" ! "first" ( "waitElves" ) . "showIn" ( emp ) ; "inofficeElves" ::= "plus" ( "inofficeElves" # "first" ( "waitElves" ) ) ; "waitElves" ::= "rest" ( "waitElves" ) ; ( await ( "t1" ?? ) ) ; ( await ( "t2" ?? ) ) ; ( await ( "t3" ?? ) ) ; "t1" ! "first" ( "inoffcieElves" ) . "showOut" ( emp ) ; "inofficeElves" ::= "rest" ( "inofficeElves" ) ; "t2" ! "first" ( "inoffcieElves" ) . "showOut" ( emp ) ; "inofficeElves" ::= "rest" ( "inofficeElves" ) ; "t3" ! "first" ( "inoffcieElves" ) . "showOut" ( emp ) ; "inofficeElves" ::= "rest" ( "inofficeElves" ) ; ( await ( "t1" ?? ) ) ; ( await ( "t2" ?? ) ) ; ( await ( "t3" ?? ) ) ; return ( emp ) > *
  < "backFromHoliday" : Mtdname | Param: noAid, Latt: noSubst, Code: "ctReindeers" ::= "+" ( "ctReindeers" # int(1) ) ; "waitReindeers" ::= "plus" ( "waitReindeers" # "caller" ) ; return ( emp ) > *
  < "haveProblem" : Mtdname | Param: noAid, Latt: noSubst, Code: "ctElves" ::= "+" ( "ctElves" # int(1) ) ; "waitElves" ::= "plus" ( "waitElves" # "caller" ) ; return ( emp ) >, Ocnt: 0 >

< "ReindeerC" : Cl | Inh: noInh, Par: "sc", Att: "sc" |-> null, Mtds: 
  < "run" : Mtdname | Param: noAid, Latt: noSubst, Code: ( ".anon" ! "this" . "holiday" (emp ) )  ; free( ".anon" ) ; return ( emp ) > *
  < "holiday" : Mtdname | Param: noAid, Latt: noSubst, Code: ".anon" ! "sc" . "backFromHoliday" ( emp ) ; free( ".anon" ) ; return ( emp ) > *
  < "deliverToys" : Mtdname | Param: noAid, Latt: noSubst, Code: skip ; return ( emp ) > *
  < "harness" : Mtdname | Param: noAid, Latt: noSubst, Code: ( ".anon" ! "this" . "deliverToys" (emp ) )  ; free( ".anon" ) ; return ( emp ) > *
  < "unharness" : Mtdname | Param: noAid, Latt: noSubst, Code: ( ".anon" ! "this" . "holiday" (emp ) )  ; free( ".anon" ) ; return ( emp ) >, Ocnt: 0 >

< "ElfC" : Cl | Inh: noInh, Par: "sc", Att: "sc" |-> null, Mtds: 
  < "run" : Mtdname | Param: noAid, Latt: noSubst, Code: ( ".anon" ! "this" . "work" (emp ) )  ; free( ".anon" ) ; return ( emp ) > *
  < "work" : Mtdname | Param: noAid, Latt: noSubst, Code: ".anon" ! "sc" . "haveProblem" ( emp ) ; free( ".anon" ) ; return ( emp ) > *
  < "talkToSanta" : Mtdname | Param: noAid, Latt: noSubst, Code: skip ; return ( emp ) > *
  < "showIn" : Mtdname | Param: noAid, Latt: noSubst, Code: ( ".anon" ! "this" . "talkToSanta" (emp ) )  ; free( ".anon" ) ; return ( emp ) > *
  < "showOut" : Mtdname | Param: noAid, Latt: noSubst, Code: ( ".anon" ! "this" . "work" (emp ) )  ; free( ".anon" ) ; return ( emp ) >, Ocnt: 0 >

< "LeaderElfC" : Cl | Inh: "ElfC" < emp > , Par: "sc" , "elves", Att: "sc" |-> null , "elves" |-> null, Mtds: 
  < "run" : Mtdname | Param: noAid, Latt: noSubst, Code: ( "sync.0" ! "run" @ "ElfC" ( emp ) )  ; ( "sync.0" ? ( noAid ) )  ; return ( emp ) > *
  < "work" : Mtdname | Param: noAid, Latt: noSubst, Code: ".anon" ! "sc" . "haveProblem" ( emp ) ; free( ".anon" ) ; return ( emp ) >, Ocnt: 0 >

< "Christmas" : Cl | Inh: noInh, Par: noAid, Att: noSubst, Mtds: 
  < "run" : Mtdname | Param: noAid, Latt: "sc" |-> null , "r" |-> null , "e" |-> null, Code: "sc" ::= new "SantaClausClass" ( emp ) ; "r" ::= new "Reindeer" ( "sc" ) ; "r" ::= new "Reindeer" ( "sc" ) ; "r" ::= new "Reindeer" ( "sc" ) ; "r" ::= new "Reindeer" ( "sc" ) ; "r" ::= new "Reindeer" ( "sc" ) ; "e" ::= new "Elf" ( "sc" ) ; "e" ::= new "Elf" ( "sc" ) ; "e" ::= new "Elf" ( "sc" ) ; "e" ::= new "Elf" ( "sc" ) ; "e" ::= new "Elf" ( "sc" ) ; "e" ::= new "Elf" ( "sc" ) ; "e" ::= new "Elf" ( "sc" ) ; return ( emp ) >, Ocnt: 0 >

.
endm
