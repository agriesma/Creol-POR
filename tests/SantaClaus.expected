==========================================
reduce in PROGRAM : init .
result [Configuration]: 
< 'Christmas : Cl | Inh: noInh,Par: noAid,Att: noSubst,Mtds:
< 'run : Mtdname | Param: noAid,Latt: 'e |-> null, 'r |-> null, 'sc |-> null,
    Code: 'sc ::= new 'SantaClausClass(emp) ; 'r ::= new 'Reindeer('sc) ; 'r
    ::= new 'Reindeer('sc) ; 'r ::= new 'Reindeer('sc) ; 'r ::= new 'Reindeer(
    'sc) ; 'r ::= new 'Reindeer('sc) ; 'e ::= new 'Elf('sc) ; 'e ::= new 'Elf(
    'sc) ; 'e ::= new 'Elf('sc) ; 'e ::= new 'Elf('sc) ; 'e ::= new 'Elf('sc) ;
    'e ::= new 'Elf('sc) ; 'e ::= new 'Elf('sc) ; return(emp) >,Ocnt: 0 >
 
< 'ElfC : Cl | Inh: noInh,Par: 'sc,Att: 'sc |-> null,Mtds:
< 'run : Mtdname | Param: noAid,Latt: noSubst,Code: ! 'this . 'work(emp) ;
    return(emp) > *
< 'showIn : Mtdname | Param: noAid,Latt: noSubst,Code: ! 'this . 'talkToSanta(
    emp) ; return(emp) > *
< 'showOut : Mtdname | Param: noAid,Latt: noSubst,Code: ! 'this . 'work(emp) ;
    return(emp) > *
< 'talkToSanta : Mtdname | Param: noAid,Latt: noSubst,Code: skip ; return(emp)
    > *
< 'work : Mtdname | Param: noAid,Latt: noSubst,Code: ! 'sc . 'haveProblem(emp)
    ; return(emp) >,Ocnt: 0 >
 
< 'LeaderElfC : Cl | Inh: 'ElfC < emp >,Par: 'sc,,'elves,Att: 'elves |-> null,
    'sc |-> null,Mtds:
< 'run : Mtdname | Param: noAid,Latt: noSubst,Code: 'run @ 'ElfC(emp : noAid) ;
    return(emp) > *
< 'work : Mtdname | Param: noAid,Latt: noSubst,Code: ! 'sc . 'haveProblem(emp)
    ; return(emp) >,Ocnt: 0 >
 
< 'ReindeerC : Cl | Inh: noInh,Par: 'sc,Att: 'sc |-> null,Mtds:
< 'deliverToys : Mtdname | Param: noAid,Latt: noSubst,Code: skip ; return(emp)
    > *
< 'harness : Mtdname | Param: noAid,Latt: noSubst,Code: ! 'this . 'deliverToys(
    emp) ; return(emp) > *
< 'holiday : Mtdname | Param: noAid,Latt: noSubst,Code: ! 'sc .
    'backFromHoliday(emp) ; return(emp) > *
< 'run : Mtdname | Param: noAid,Latt: noSubst,Code: ! 'this . 'holiday(emp) ;
    return(emp) > *
< 'unharness : Mtdname | Param: noAid,Latt: noSubst,Code: ! 'this . 'holiday(
    emp) ; return(emp) >,Ocnt: 0 >
 
< 'SantaClausC : Cl | Inh: noInh,Par: noAid,Att: 'ctElves |-> null,
    'ctRreindeers |-> null, 'harnessedReindeers |-> null, 'inofficeElves |->
    null, 'waitElves |-> null, 'waitRreindeers |-> null,Mtds:
< 'backFromHoliday : Mtdname | Param: noAid,Latt: noSubst,Code: 'ctReindeers
    ::= ('plus[['ctReindeers # int(1)]]) ; 'waitReindeers ::= ('plus[[
    'waitReindeers # 'caller]]) ; return(emp) > *
< 'deliverToys : Mtdname | Param: noAid,Latt: noSubst,Code: 'ctReindeers ::=
    int(0) ; 't1 ! ('first[['waitReindeers]]) . 'harness(emp) ;
    'harnessedReindeers ::= ('plus[['harnessedReindeers # ('first[[
    'waitReindeers]])]]) ; 'waitReindeers ::= ('rest[['waitReindeers]]) ; 't5 !
    ('first[['waitReindeers]]) . 'harness(emp) ; 'harnessedReindeers ::= (
    'plus[['harnessedReindeers # ('first[['waitReindeers]])]]) ; 'waitReindeers
    ::= ('rest[['waitReindeers]]) ; await ('t1 ? 'and 't2 ? 'and 't3 ? 'and 't4
    ? 'and 't5 ?) ; ! ('first[['harnessedReindeers]]) . 'unharness(emp) ;
    'harnessedReindeers ::= ('rest[['harnessedReindeers]]) ; ! ('first[[
    'harnessedReindeers]]) . 'unharness(emp) ; 'harnessedReindeers ::= ('rest[[
    'harnessedReindeers]]) ; return(emp) > *
< 'haveProblem : Mtdname | Param: noAid,Latt: noSubst,Code: 'ctElves ::= (
    'plus[['ctElves # int(1)]]) ; 'waitElves ::= ('plus[['waitElves #
    'caller]]) ; return(emp) > *
< 'loop : Mtdname | Param: noAid,Latt: noSubst,Code: (await ('equal[[
    'ctReindeers # int(5)]]) ; 'deliverToys(emp : noAid)[]await ('equal[[('and[
    [('lessEq[[int(3) # 'ctElves]]) # ('not[['ctReindeers]])]]) # int(9)]]) ;
    'talkToElves(emp : noAid)) ; ! 'this . 'loop(emp) ; return(emp) > *
< 'run : Mtdname | Param: noAid,Latt: noSubst,Code: ! 'this . 'loop(emp) ;
    return(emp) > *
< 'talkToelves : Mtdname | Param: noAid,Latt: noSubst,Code: 'ctElves ::= (
    'minus[['ctElves # int(3)]]) ; 't1 ! ('first[['waitElves]]) . 'showIn(emp)
    ; 'inofficeElves ::= ('plus[['inofficeElves # ('first[['waitElves]])]]) ;
    'waitElves ::= ('rest[['waitElves]]) ; 't2 ! ('first[['waitElves]]) .
    'showIn(emp) ; 'inofficeElves ::= ('plus[['inofficeElves # ('first[[
    'waitElves]])]]) ; 'waitElves ::= ('rest[['waitElves]]) ; 't3 ! ('first[[
    'waitElves]]) . 'showIn(emp) ; 'inofficeElves ::= ('plus[['inofficeElves #
    ('first[['waitElves]])]]) ; 'waitElves ::= ('rest[['waitElves]]) ; await (
    't1 ? 'and 't2 ? 'and 't3 ?) ; 't1 ! ('first[['inoffcieElves]]) . 'showOut(
    emp) ; 'inofficeElves ::= ('rest[['inofficeElves]]) ; 't2 ! ('first[[
    'inoffcieElves]]) . 'showOut(emp) ; 'inofficeElves ::= ('rest[[
    'inofficeElves]]) ; 't3 ! ('first[['inoffcieElves]]) . 'showOut(emp) ;
    'inofficeElves ::= ('rest[['inofficeElves]]) ; await ('t1 ? 'and 't2 ? 'and
    't3 ?) ; return(emp) >,Ocnt: 0 >

Bye.
